File: avalon-modules/avalon-dataset/avalon-dataset-service/src/test/java/io/github/lishangbu/avalon/dataset/service/impl/StatServiceImplTest.java
package io.github.lishangbu.avalon.dataset.service.impl;

import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertSame;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import io.github.lishangbu.avalon.dataset.entity.Stat;
import io.github.lishangbu.avalon.dataset.repository.StatRepository;
import java.util.List;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.data.domain.Example;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;

@ExtendWith(MockitoExtension.class)
class StatServiceImplTest {
    @Mock private StatRepository statRepository;
    @InjectMocks private StatServiceImpl statService;

    @Test
    void getPageByCondition_callsRepository() {
        Stat stat = new Stat();
        Pageable pageable = PageRequest.of(0, 5);
        Page<Stat> page = new PageImpl<>(List.of(stat));
        when(statRepository.findAll(any(Example.class), eq(pageable))).thenReturn(page);

        Page<Stat> result = statService.getPageByCondition(stat, pageable);

        assertSame(page, result);
        ArgumentCaptor<Example> captor = ArgumentCaptor.forClass(Example.class);
        verify(statRepository).findAll(captor.capture(), eq(pageable));
        assertNotNull(captor.getValue().getProbe());
    }

    @Test
    void listByCondition_callsRepository() {
        Stat stat = new Stat();
        List<Stat> expected = List.of(stat);
        when(statRepository.findAll(any(Example.class))).thenReturn(expected);

        List<Stat> result = statService.listByCondition(stat);

        assertSame(expected, result);
        verify(statRepository).findAll(any(Example.class));
    }

    @Test
    void save_usesRepository() {
        Stat stat = new Stat();
        when(statRepository.save(stat)).thenReturn(stat);

        Stat result = statService.save(stat);

        assertSame(stat, result);
        verify(statRepository).save(stat);
    }

    @Test
    void update_usesRepository() {
        Stat stat = new Stat();
        when(statRepository.save(stat)).thenReturn(stat);

        Stat result = statService.update(stat);

        assertSame(stat, result);
        verify(statRepository).save(stat);
    }

    @Test
    void removeById_callsRepository() {
        statService.removeById(1L);
        verify(statRepository).deleteById(1L);
    }
}

File: avalon-modules/avalon-dataset/avalon-dataset-service/src/test/kotlin/io/github/lishangbu/avalon/dataset/service/impl/StatServiceImplTest.kt
package io.github.lishangbu.avalon.dataset.service.impl

import io.github.lishangbu.avalon.dataset.entity.MoveDamageClass
import io.github.lishangbu.avalon.dataset.entity.Stat
import io.github.lishangbu.avalon.dataset.entity.dto.SaveStatInput
import io.github.lishangbu.avalon.dataset.entity.dto.StatSpecification
import io.github.lishangbu.avalon.dataset.entity.dto.StatView
import io.github.lishangbu.avalon.dataset.entity.dto.UpdateStatInput
import io.github.lishangbu.avalon.dataset.repository.StatRepository
import org.babyfish.jimmer.sql.ast.mutation.AssociatedSaveMode
import org.babyfish.jimmer.sql.ast.mutation.SaveMode
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Test
import org.mockito.Mockito.mock
import org.mockito.Mockito.verify
import org.mockito.Mockito.`when`

class StatServiceImplTest {
    private val repository = mock(StatRepository::class.java)
    private val service = StatServiceImpl(repository)

    @Test
    fun listByCondition_callsRepository() {
        val specification = StatSpecification(id = "1", internalName = "speed")
        `when`(repository.listViews(specification)).thenReturn(
            listOf(
                StatView(
                    "2",
                    "attack",
                    "攻击",
                    2,
                    false,
                    false,
                    StatView.TargetOf_moveDamageClass("2", "physical", "物理"),
                ),
            ),
        )

        val result = service.listByCondition(specification)

        assertEquals(1, result.size)
        assertEquals("2", result.first().id)
        assertEquals("physical", result.first().moveDamageClass?.internalName)
    }

    @Test
    fun save_usesInsertOnlyModeAndReloadsView() {
        `when`(repository.save(any<Stat>(), eq(SaveMode.INSERT_ONLY), eq(AssociatedSaveMode.REPLACE), isNull())).thenReturn(statSavedEntity(2L))
        `when`(repository.loadViewById(2L)).thenReturn(statView("2", "attack", "攻击"))

        val result = service.save(SaveStatInput("attack", "攻击", 2, false, false, "2"))

        assertEquals("2", result.id)
        assertEquals("物理", result.moveDamageClass?.name)
        verify(repository).save(any<Stat>(), eq(SaveMode.INSERT_ONLY), eq(AssociatedSaveMode.REPLACE), isNull())
        verify(repository).loadViewById(2L)
    }

    @Test
    fun update_usesUpsertModeAndReloadsView() {
        `when`(repository.findNullable(2L)).thenReturn(statSavedEntity(2L))
        `when`(repository.save(any<Stat>(), eq(SaveMode.UPDATE_ONLY), eq(AssociatedSaveMode.REPLACE), isNull())).thenReturn(statSavedEntity(2L))
        `when`(repository.loadViewById(2L)).thenReturn(statView("2", "attack", "攻击"))

        val result = service.update(UpdateStatInput("2", "attack", "攻击", 2, false, false, "2"))

        assertEquals("2", result.id)
        assertEquals("physical", result.moveDamageClass?.internalName)
        verify(repository).save(any<Stat>(), eq(SaveMode.UPDATE_ONLY), eq(AssociatedSaveMode.REPLACE), isNull())
        verify(repository).loadViewById(2L)
    }

    @Test
    fun removeById_callsRepository() {
        `when`(repository.findNullable(1L)).thenReturn(statSavedEntity(1L))

        service.removeById(1L)

        verify(repository).deleteById(1L)
    }

    @Test
    fun update_rejectsReadonlyStat() {
        `when`(repository.findNullable(2L)).thenReturn(statSavedEntity(2L, readonly = true))

        val error =
            assertThrows(IllegalStateException::class.java) {
                service.update(UpdateStatInput("2", "attack", "攻击", 2, false, true, "2"))
            }

        assertEquals("能力值已设为只读，禁止修改", error.message)
    }

    @Test
    fun removeById_rejectsReadonlyStat() {
        `when`(repository.findNullable(1L)).thenReturn(statSavedEntity(1L, readonly = true))

        val error =
            assertThrows(IllegalStateException::class.java) {
                service.removeById(1L)
            }

        assertEquals("能力值已设为只读，禁止删除", error.message)
    }
}

private fun statSavedEntity(
    id: Long,
    readonly: Boolean = false,
): Stat =
    Stat {
        this.id = id
        internalName = "attack"
        name = "攻击"
        gameIndex = 2
        battleOnly = false
        this.readonly = readonly
        moveDamageClass =
            MoveDamageClass {
                this.id = 2L
            }
    }

private fun statView(
    id: String,
    internalName: String,
    name: String,
    readonly: Boolean = false,
): StatView =
    StatView(
        id,
        internalName,
        name,
        2,
        false,
        readonly,
        StatView.TargetOf_moveDamageClass("2", "physical", "物理"),
    )
